src/storage/src/upsert_continual_feedback.rs

```diff
@@ -246,170 +275,115 @@ where
-                        };
-
-                        tracing::debug!(
-                            worker_id = %source_config.worker_id,
-                            source_id = %source_config.id,
-                            persist_stash = %persist_stash.len(),
-                            %last_snapshot_chunk,
-                            ?resume_upper,
-                            ?persist_upper,
-                            "ingesting persist snapshot chunk");
-
-                        hydrating = !last_snapshot_chunk;
-
-                        let ready_updates = persist_stash.drain_filter_swapping(|(_key, value, ts, diff)| {
-                            let first_source_snapshot = ts == &G::Timestamp::minimum();
-                            let rehydration_snapshot = !resume_upper.less_equal(ts);
-                            let ready = first_source_snapshot || rehydration_snapshot;
-
-                            // Also collect any legacy errors as they're flying
-                            // by.
-                            if ready {
-                                if let Err(UpsertError::Value(ref mut err)) = value {
-                                    // If we receive a legacy error in the snapshot we will keep a note of it but
-                                    // insert a non-legacy error in our state. This is so that if this error is
-                                    // ever retracted we will correctly retract the non-legacy version because by
-                                    // that time we will have emitted the error correction, which happens before
-                                    // processing any of the new source input.
-                                    if err.is_legacy_dont_touch_it {
-                                        legacy_errors_to_correct.push((err.clone(), diff.clone()));
-                                        err.is_legacy_dont_touch_it = false;
-                                    }
-                                }
+                    // When we finish ingesting our initial persist snapshot,
+                    // during "re-hydration", we downgrade this to the empty
+                    // frontier, so we need to be lenient to this failing from
+                    // then on.
+                    let _ = snapshot_cap.try_downgrade(persist_upper.iter());
+
+                    let last_rehydration_chunk =
+                        hydrating && PartialOrder::less_equal(&resume_upper, &persist_upper);
+
+                    tracing::debug!(
+                        worker_id = %source_config.worker_id,
+                        source_id = %source_config.id,
+                        persist_stash = %persist_stash.len(),
+                        %last_rehydration_chunk,
+                        ?resume_upper,
+                        ?persist_upper,
+                        "ingesting persist snapshot chunk");
+
+                    let ready_updates = persist_stash.drain(..);
```

Comment:
What does ready mean here? I see that we use the entirety of `persist_stash` but that might contain multiple timestamps, including timestamps that are beyond the current upper (see comment on `ingesting_persist_timestamps`

Reply:
The name is a leftover from a previous state of the code where I wasn't always ingesting all updates. Now we eagerly ingest everything, and make sure to only touch state once we know it's in a valid state (by using `ingesting_persist_timestamps`).